### PR TITLE
Add interactive UI example Snap

### PR DESCRIPTION
**Pull request by GuillaumeRx**
This PR adds a new example snap to demonstrate the use of interactive UI across the various supported APIs and creates its own section in `test-snaps`.

Closes https://github.com/MetaMask/MetaMask-planning/issues/2039

### CODE REVIEW

File: packages/examples/packages/interactive-ui/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+  text,
+  copyable,
+} from '@metamask/snaps-sdk';
+
+/**
+ * Initiate a new interface with the starting screen and return the Snap interface ID.
+ *
+ * @returns The Snap interface ID.
+ */
+export const createInterface = async () =>
```
Comment (hmalik88) [inline packages/examples/packages/interactive-ui/src/index.ts:25]:
```suggestion
export const createInterface = async (): string =>
```
nit

File: packages/examples/packages/interactive-ui/CHANGELOG.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+## [Unreleased]
+
+## [1.0.0]
+
+### Added
+
+// TODO: Change when we have the PR number
+
+- Add interactive UI example snap ([#2159](https://github.com/MetaMask/snaps/pull/xxxx))
```
Comment (FrederikBolding) [inline packages/examples/packages/interactive-ui/CHANGELOG.md:16]: Just leave this empty instead

File: packages/examples/packages/interactive-ui/src/index.test.ts
Comment (hmalik88) [inline packages/examples/packages/interactive-ui/src/index.test.ts:1]: Needs to be updated

File: packages/examples/packages/interactive-ui/CHANGELOG.md [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
+
+## [Unreleased]
+
+## [1.0.0]
+
+### Added
+
+[Unreleased]: https://github.com/MetaMask/snaps/compare/@metamask/interactive-ui-example-snap@1.0.0...HEAD
+[1.0.0]: https://github.com/MetaMask/snaps/releases/tag/@metamask/interactive-ui-example-snap@1.0.0
```
Comment (FrederikBolding) [inline packages/examples/packages/interactive-ui/CHANGELOG.md:15]:
```suggestion

[Unreleased]: https://github.com/MetaMask/snaps/
```

File: packages/examples/packages/interactive-ui/README.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
+# `@metamask/interactive-ui-example-snap`
+
+This snap demonstrates how to use interactive UI to build reactive custom UI interfaces accross all the available APIs.
```
Comment (FrederikBolding) [inline packages/examples/packages/interactive-ui/README.md:3]:
```suggestion
This snap demonstrates how to use interactive UI to build reactive custom UI interfaces across all the available APIs.
```

File: packages/test-snaps/src/features/snaps/index.ts [hunk 5]
```diff
@@ -20,3 +20,4 @@ export * from './transaction-insights';
 export * from './signature-insights';
 export * from './updates';
 export * from './wasm';
+export * from './interactive-ui';
```
Comment (FrederikBolding) [inline packages/test-snaps/src/features/snaps/index.ts:23]: Please sort alphabetically

File: packages/examples/packages/interactive-ui/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+ * @see https://docs.metamask.io/snaps/reference/rpc-api/#wallet_invokesnap
+ */
+export const onRpcRequest: OnRpcRequestHandler<{ id: string }> = async ({
+  request,
+}) => {
+  switch (request.method) {
+    case 'dialog': {
+      const interfaceId = await createInterface();
+
+      snap.request({
```
Comment (FrederikBolding) [inline packages/examples/packages/interactive-ui/src/index.ts:103]: Will the snap not terminate by being idle if you do this instead of awaiting?

**Review by Mrtenz — CHANGES_REQUESTED** [review]
Please update the `README.md` in the examples folder too.

File: packages/examples/packages/interactive-ui/README.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+## Snap usage
+
+### onRpcRequest
+
+This snap exposes an `onRpcRequest` handler, which supports the following
+JSON-RPC methods:
+
+- `dialog`: Create a `snap_dialog` with an interactive interface. This demonstrates that a snap can show an interactive `snap_dialog` that the user can interact with.
+
+- `get_state`: Get the state of a given interface. This demonstrates that a snap can retrieve an interface state.
```
Comment (Mrtenz) [inline packages/examples/packages/interactive-ui/README.md:14]:
```suggestion
- `getState`: Get the state of a given interface. This demonstrates that a snap can retrieve an interface state.
```

File: packages/examples/packages/interactive-ui/README.md [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+### onTransaction
+
+This snap exposes an `onTransaction` handler, which is called when a transaction
+is sent by the user.
+
+The snap creates a new Snap interface and returns the ID.
+
+### onHomePage
+
+This snap exposes an `onHomePage` handler, which creates a new Snap interface and returns the ID to be shown.
```
Comment (Mrtenz) [inline packages/examples/packages/interactive-ui/README.md:25]:
```suggestion
The snap exposes an `onHomePage` handler, which shows a user interface.
```

File: packages/examples/packages/interactive-ui/README.md [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+- `dialog`: Create a `snap_dialog` with an interactive interface. This demonstrates that a snap can show an interactive `snap_dialog` that the user can interact with.
+
+- `get_state`: Get the state of a given interface. This demonstrates that a snap can retrieve an interface state.
+
+### onTransaction
+
+This snap exposes an `onTransaction` handler, which is called when a transaction
+is sent by the user.
+
+The snap creates a new Snap interface and returns the ID.
```
Comment (Mrtenz) [inline packages/examples/packages/interactive-ui/README.md:21]:
```suggestion
This snap exposes an `onTransaction` handler, which is called when a transaction
is sent by the user. It shows a user interface with details about the transaction.
```

File: packages/examples/packages/interactive-ui/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+  text,
+  copyable,
+} from '@metamask/snaps-sdk';
+
+/**
+ * Initiate a new interface with the starting screen and return the Snap interface ID.
+ *
+ * @returns The Snap interface ID.
+ */
+export const createInterface = async (): Promise<string> =>
```
Comment (Mrtenz) [inline packages/examples/packages/interactive-ui/src/index.ts:25]: We usually use regular functions instead of arrow functions for these things.

Then you can use `return await`.

File: packages/examples/packages/interactive-ui/src/index.ts [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+ * transaction is signed, so it can be used to display information about the
+ * transaction to the user before they sign it.
+ *
+ * The `onTransaction` handler returns a Snaps interface ID, which is displayed
+ * in the transaction insights panel.
+ *
+ * @returns The transaction insights.
+ */
+export const onTransaction: OnTransactionHandler = async () => {
+  const interfaceId = await createInterface();
```
Comment (Mrtenz) [inline packages/examples/packages/interactive-ui/src/index.ts:172]: Maybe create a slightly different UI for each entry point, to demonstrate how to do that? This could display some info about the transaction for example.

File: packages/examples/packages/interactive-ui/src/index.ts [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+ * method. This handler decodes the transaction data, and displays the type of
+ * transaction in the transaction insights panel.
+ *
+ * The `onTransaction` handler is different from the `onRpcRequest` handler in
+ * that it is called by MetaMask when a transaction is initiated, rather than
+ * when a dapp sends a JSON-RPC request. The handler is called before the
+ * transaction is signed, so it can be used to display information about the
+ * transaction to the user before they sign it.
+ *
+ * The `onTransaction` handler returns a Snaps interface ID, which is displayed
```
Comment (Mrtenz) [inline packages/examples/packages/interactive-ui/src/index.ts:166]: It's not the ID that's being displayed. I would rephrase this a bit.

File: packages/examples/packages/interactive-ui/src/index.ts (hunk 6 above)
Comment (GuillaumeRx) [reply packages/examples/packages/interactive-ui/src/index.ts:103]: I don't know but I can't wait for the confirmation to resolve to return the interface ID, I need it while the confirmation is openend. I could store the interface ID and have another method to get it.

Comment (FrederikBolding) [reply packages/examples/packages/interactive-ui/src/index.ts:103]: Yeah, right now this snap would be consider idle because we only consider inbound requests when determining if the snap is idle or not.

File: packages/examples/packages/interactive-ui/src/index.ts [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+};
+
+/**
+ * Update a Snap interface to show a given value.
+ *
+ * @param id - The Snap interface ID to update.
+ * @param value - The value to display in the UI.
+ */
+export const showResult = async (id: string, value: string) => {
+  snap.request({
```
Comment (FrederikBolding) [inline packages/examples/packages/interactive-ui/src/index.ts:67]: Should we not await this?

File: packages/examples/packages/interactive-ui/src/index.ts [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+    },
+  });
+
+/**
+ * Update the interface with a simple form containing an input and a submit button.
+ *
+ * @param id - The Snap interface ID to update.
+ */
+export const showForm = async (id: string) => {
+  snap.request({
```
Comment (FrederikBolding) [inline packages/examples/packages/interactive-ui/src/index.ts:42]: Should we not await this?

File: packages/examples/packages/interactive-ui/src/index.ts [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+  copyable,
+} from '@metamask/snaps-sdk';
+
+/**
+ * Initiate a new interface with the starting screen and return the Snap interface ID.
+ *
+ * @returns The Snap interface ID.
+ */
+export const createInterface = async (): Promise<string> =>
+  snap.request({
```
Comment (FrederikBolding) [inline packages/examples/packages/interactive-ui/src/index.ts:26]: Should we not await this?

File: packages/examples/packages/interactive-ui/CHANGELOG.md (hunk 3 above)
Comment (FrederikBolding) [reply packages/examples/packages/interactive-ui/CHANGELOG.md:15]: This has not been resolved

File: packages/examples/packages/interactive-ui/images/icon.svg
Comment (FrederikBolding) [inline packages/examples/packages/interactive-ui/images/icon.svg:1]: Delete please!

File: packages/examples/packages/interactive-ui/package.json [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+    "type": "git",
+    "url": "https://github.com/MetaMask/snaps.git"
+  },
+  "license": "(MIT-0 OR Apache-2.0)",
+  "sideEffects": false,
+  "main": "./dist/bundle.js",
+  "files": [
+    "dist/",
+    "snap.manifest.json",
+    "images/icon.svg"
```
Comment (FrederikBolding) [inline packages/examples/packages/interactive-ui/package.json:15]:
```suggestion
```

File: packages/examples/packages/interactive-ui/snap.manifest.json [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/MetaMask/snaps.git"
+  },
+  "source": {
+    "shasum": "IAAEJ2BZ7JLp1KFTYTJmyOhiGQ0dxMEQk/pb0p+ffhU=",
+    "location": {
+      "npm": {
+        "filePath": "dist/bundle.js",
+        "iconPath": "images/icon.svg",
```
Comment (FrederikBolding) [inline packages/examples/packages/interactive-ui/snap.manifest.json:14]:
```suggestion
```

File: packages/examples/packages/interactive-ui/src/index.ts [hunk 18]
```diff
@@ -0,0 +1,331 @@
...
+  ]);
+}
+
+/**
+ * Updates a Snap interface to display the transaction type.
+ *
+ * @param id -  The interface ID to update.
+ */
+export async function displayTransactionType(id: string) {
+  let type = 'unkown';
```
Comment (FrederikBolding) [inline packages/examples/packages/interactive-ui/src/index.ts:76]:
```suggestion
  let type = 'unknown';
```

File: packages/examples/packages/interactive-ui/src/index.ts [hunk 19]
```diff
@@ -0,0 +1,331 @@
...
+import { hasProperty } from '@metamask/utils';
+
+import { decodeData } from './utils';
+
+/**
+ * Initiate a new interface with the starting screen.
+ *
+ * @returns The Snap interface ID.
+ */
+export async function createInterface(): Promise<string> {
```
Comment (FrederikBolding) [inline packages/examples/packages/interactive-ui/src/index.ts:34]: Maybe it would be a little bit simpler and easier to read if we move some of these to a `ui` file. WDYT?

File: packages/examples/packages/interactive-ui/src/index.ts [hunk 20]
```diff
@@ -0,0 +1,331 @@
...
+import type {
+  OnHomePageHandler,
+  OnTransactionHandler,
+  OnUserInputHandler,
+  Transaction,
+  Component,
+} from '@metamask/snaps-sdk';
+import {
+  panel,
+  type OnRpcRequestHandler,
```
Comment (Mrtenz) [inline packages/examples/packages/interactive-ui/src/index.ts:11]: Can be moved to the type import above.

File: packages/examples/packages/interactive-ui/src/index.ts [hunk 21]
```diff
@@ -0,0 +1,331 @@
...
+
+  return panel([
+    row('From', address(from)),
+    row('To', to ? address(to) : text('None')),
+    button({ value: 'See transaction type', name: 'transaction-type' }),
+  ]);
+}
+
+/**
+ * Updates a Snap interface to display the transaction type.
```
Comment (Mrtenz) [inline packages/examples/packages/interactive-ui/src/index.ts:71]:
```suggestion
 * Update a Snap interface to display the transaction type.
```

File: packages/examples/packages/interactive-ui/src/index.ts [hunk 22]
```diff
@@ -0,0 +1,331 @@
...
+ * @param params.event - The event object containing the event type, name and value.
+ * @see https://docs.metamask.io/snaps/reference/exports/#onuserinput
+ */
+export const onUserInput: OnUserInputHandler = async ({ id, event }) => {
+  if (
+    event.type === UserInputEventType.ButtonClickEvent &&
+    event.name === 'update'
+  ) {
+    await showForm(id);
+    return;
```
Comment (Mrtenz) [inline packages/examples/packages/interactive-ui/src/index.ts:299]: `return await ...` is a bit cleaner imo.

File: packages/examples/packages/interactive-ui/src/index.ts (hunk 18 above)
Comment (Mrtenz) [inline packages/examples/packages/interactive-ui/src/index.ts:76]: Maybe change the `decodeData` function to support `undefined` instead of using mutable variables.

File: packages/examples/packages/interactive-ui/src/index.ts (hunk 21 above)
Comment (Mrtenz) [reply packages/examples/packages/interactive-ui/src/index.ts:71]: Maybe add an explanation too that the transaction is stored in the state.

**Review by Mrtenz — CHANGES_REQUESTED** [review]

**Review by FrederikBolding — APPROVED** [review]
LGTM, but please give @Mrtenz a chance to review as well!

**Review by Mrtenz — APPROVED** [review]
